Extract ExternalCommand and CleanedOutput classes

chrismytton opened the pull request:
This adds two new classes, `ExternalCommand`, which runs external commands and captures their output, and `CleanedOutput`, which cleans up the output from external commands ready for inclusion in a pull request.

Fixes #43 
Fixes #45 
Part of #62

chrismytton committed:
Add some README details about ExternalCommand class

This outlines the expected usage for the ExternalCommand class.

chrismytton committed:
Add ExternalCommand class

This class encapsulates running an external command and gathering up the
output.

tmtmtmtm commented on README.md:
```diff
@@ -36,3 +36,24 @@ Run the app server
+
+The main task this app performs is running the EveryPolitician build process as an external command and then creating a pull request with the output from the command. As such, one of the main classes in the system is the `ExternalCommand` class. This encapsulates running an external command, checking for errors and returning the output.
+
+To use this class create an instance of `ExternalCommand` and pass `command` and `end` keyword arguments to the constructor. The `command` should be a string representing the external command to execute, and `env` is an optional hash containing environment variables that should be set, or if they are `nil`, unset.
```
typo: end → env

chrismytton committed:
Add CleanedOutput class

This class is responsible for taking the output from an external command
and cleaning it up ready to be included in a GitHub comment.

chrismytton committed:
Switch app.rb to use ExternalCommand and CleanedOutput classes

This simplifies the `RebuilderJob#perform` method quite a bit, and means
that we're using tested classes, rather than just having a big ball of
code.

chrismytton committed:
Don't mess with ENV in the tests

Rather than needing to override ENV we instead allow passing in a Morph
API key, and then defaulting that to an environment variable.

chrismytton committed:
Make sure output is correct when no Morph API key provided

Previously this was accidentally replacing every character with REDACTED
when the API key was "".

chrismytton committed:
Add details to README about CleanedOutput class

chrismytton commented:
Oops, requested a review before I'd updated the title and body. Fixed now.

tmtmtmtm commented on app.rb:
```diff
@@ -49,36 +50,28 @@ def perform(country_slug, legislature_slug, source = nil)
     message = "#{country.name}: Refresh from upstream changes"
-    output = ''
-    child_status = nil
+    command = nil
```
This raises a few warning signs to me. In general, assigning a placeholder that you're going to re-assign later is quite suspicious, and here's not really obvious why it's needed. 

I think it's also a little badly named, as it's named for what it _is_ (which is obvious as it's going be an ExternalCommand object), rather than what it's doing (i.e. it's like naming your variables `integer` or `string`), and leads to some confusion when it's later used again at line 68, even though there's been several other commands executed in the meantime. 

Perhaps it would be clearer and simpler to hoist the declaration of this command (if not others) out of within the `with_git_repo` blocks, and then only trigger the `.run` inside?

tmtmtmtm commented on lib/cleaned_output.rb:
```diff
@@ -0,0 +1,26 @@
+require 'colorize'
+
+class CleanedOutput
+  def initialize(output:, morph_api_key: ENV['MORPH_API_KEY'])
```
it feels a little odd to me to have this class so tightly coupled to Morph-specific concepts like this, or for a class like this to be peering into ENV. This seems like it would be a lot more extensible if the argument here was something like `redact: *strings`, for a list of things you want redacted in the output.

tmtmtmtm commented on lib/cleaned_output.rb:
```diff
@@ -0,0 +1,26 @@
+
+  attr_reader :output, :morph_api_key
+
+  def cleaned_output
```
this is overloading what 'cleaned' means in a slightly odd way, where `cleaned_output` here is something different to the `CleanedOutput` of the class. This is really just the original output with redactions applied.

tmtmtmtm commented on lib/cleaned_output.rb:
```diff
@@ -0,0 +1,26 @@
+  end
+
+  def to_s
+    (cleaned_output[-64_000..-1] || cleaned_output).uncolorize
```
I suspect this would be a little clearer if the truncation code were pulled out into its own method.

tmtmtmtm replied:
The magic number here should probably be hoisted out. (We'd probably want to take the size as an optional constructor argument too, but I'm OK with leaving that part to a later enhancement if it's not as trivial as it seems it should be)

tmtmtmtm reviewed (changes requested):
Yay! This is definitely a lot better. A few interface/implementation notes in-place

chrismytton commented:
This pull request has got a bit large for my liking. I've broken off the `CleanedOutput` change into https://github.com/everypolitician/rebuilder/pull/64 so that can be reviewed separately. That pull request addresses the feedback on the `CleanedOutput` class from above.

chrismytton commented:
https://github.com/everypolitician/rebuilder/pull/64 has now been merge into `master`, that contains the output cleaner part of this PR. So the next step is to pull the external command bits into another branch off `master`.

chrismytton commented:
`CleanedOutput` work was done in #64 and #65. The `ExternalCommand` work is being done in https://github.com/everypolitician/rebuilder/pull/72. So this PR is now redundant.
